### PR TITLE
Add support for CILogon deployments in different regions than the US

### PR DESCRIPTION
**Pull request by uwwint**
Add support for CILogon deployments in different regions than the US (Australia for example) by reading the .well-defined/openid-config instead of hard coding the urls.

This patch is minimally invasive and contains backwards compability with existing CILogon configurations.

(Please replace this header with a description of your pull request. Please include *BOTH* what you did and why you made the changes. The "why" may simply be citing a relevant Galaxy issue.)
(If fixing a bug, please add any relevant error or traceback)
(For UI components, it is recommended to include screenshots or screencasts)

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:
1. Setup Galaxy with OIDC to CILogon
2. Try to login see if it works

These changes have been tested with The Australian CILogon deployment, which is under the domain: https://test.cilogon.aaf.edu.au

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

File: lib/galaxy/authnz/custos_authnz.py [hunk 1]
```diff
@@ -436,6 +436,13 @@ def _get_well_known_uri_from_url(self, provider):
             # Remove potential trailing slash to avoid "//realms"
             base_url = base_url if base_url[-1] != "/" else base_url[:-1]
             return f"{base_url}/.well-known/openid-configuration"
+        if provider == "cilogon":
+            base_url = self.config["url"]
+            # backwards compatibility. CILogon URL is given with /authorize in the examples. not sure if
+            # this applies to the wild, but let's be safe here and remove the /authorize if it exists
+            # which will lead to the correct openid configuration
+            base_url = base_url if base_url.split("/")[-1] != "authorize" else "/".join(base_url.split("/")[:-1])
```
Comment (bgruening) [inline lib/galaxy/authnz/custos_authnz.py:444]: Should we inform the admin when this happens via a log message?

Comment (uwwint) [reply lib/galaxy/authnz/custos_authnz.py:444]: I believe that is a philosophical question on what the URL parameter in openid_backends_config should represent ;)

for Custos & Keycloak it is the base url, not the authorize endpoint https://github.com/galaxyproject/galaxy/blob/77632581bc0a8a21b9279af4b86e793956edbbeb/lib/galaxy/config/sample/oidc_backends_config.xml.sample#L94 https://github.com/galaxyproject/galaxy/blob/77632581bc0a8a21b9279af4b86e793956edbbeb/lib/galaxy/config/sample/oidc_backends_config.xml.sample#L133C1-L134C1

For CILogon it is the authorize endpoint:
https://github.com/galaxyproject/galaxy/blob/77632581bc0a8a21b9279af4b86e793956edbbeb/lib/galaxy/config/sample/oidc_backends_config.xml.sample#L114C19-L114C20

Ideally there shouldn't be a difference in what the URL parameter points to between providers. If you are planning to homogenise CILogon/Custos/Keycloak down the track we could throw a message and update the template.

File: lib/galaxy/authnz/custos_authnz.py [hunk 2]
```diff
@@ -392,9 +392,9 @@ def _validate_nonce(self, trans, nonce_hash):
 
     def _load_config_for_cilogon(self):
         # Set cilogon endpoints
-        self.config["authorization_endpoint"] = "https://cilogon.org/authorize"
-        self.config["token_endpoint"] = "https://cilogon.org/oauth2/token"
-        self.config["userinfo_endpoint"] = "https://cilogon.org/oauth2/userinfo"
+        self.config["well_known_oidc_config_uri"] = self._get_well_known_uri_from_url(self.config["provider"])
```
Comment (bernt-matthias) [inline lib/galaxy/authnz/custos_authnz.py:395]: Could you extend the sample config file accordingly? Or is this already covered?

File: lib/galaxy/authnz/custos_authnz.py [hunk 3]
```diff
@@ -392,9 +392,9 @@ def _validate_nonce(self, trans, nonce_hash):
 
     def _load_config_for_cilogon(self):
```
Comment (bernt-matthias) [inline lib/galaxy/authnz/custos_authnz.py:393]: Seems that this function is now identical to `_load_config_for_keycloak(self):`. Maybe a chance to reduce code duplication?

File: lib/galaxy/authnz/custos_authnz.py (hunk 2 above)
Comment (uwwint) [reply lib/galaxy/authnz/custos_authnz.py:395]: It's already covered.

File: lib/galaxy/authnz/custos_authnz.py (hunk 3 above)
Comment (uwwint) [reply lib/galaxy/authnz/custos_authnz.py:393]: for refactoring, please see here: 
https://github.com/galaxyproject/galaxy/pull/16497

**Comment by uwwint** [comment]
Can we merge this bugfix irrespective of the larger refactoring linked?

**Comment by nuwang** [comment]
> Can we merge this bugfix irrespective of the larger refactoring linked?

This will come out in the 23.2 release. If you need it for an earlier release as a bug fix, you will need to change the base branch and retarget that branch. If not, we can merge this straightaway.

**Review by nuwang — APPROVED** [review]

**Pull request merged**